Fix module support for oneapi compilers

rscohn2 opened the pull request:
CC/CXX/FC/F77 will be set in the run environment and the module file. This makes the behavior consistent with other spack packages that include compilers: intel-parallel-studio, dpcpp, nvhpc, pgi.

Also fixes lmod problem identified in #30608.

Split out module support for icc by creating a new package intel-oneapi-compilers-classic

Users of icx can use intel-oneapi-compilers, users of icc use intel-oneapi-compilers-classic

Both packages install both compilers because the intel installer that we depend upon always installs both. intel-oneapi-compilers-classic depends on intel-oneapi-compilers and does not install anything additional.
```

rscohn2 committed:
set CC & friends for intel-oneapi-compilers

rscohn2 commented:
@scheibelp: This partially addresses https://github.com/spack/spack/issues/28768. Can you review it?

scheibelp commented:
Thanks for this, I'll be able to take a look at it by Wednesday

scheibelp commented on var/spack/repos/builtin/packages/intel-oneapi-compilers/package.py:
```diff
@@ -83,6 +83,10 @@ class IntelOneapiCompilers(IntelOneApiPackage):
 
+    variant('CC', default='icx',
+            values=('icx', 'icc', 'dpcpp'), multi=False,
+            description='Set CC & friends in run environment')
```
There's a conflict of outcomes here. On the one hand, we want to create module files that set CC etc. for convenience; to do that this PR adds a CC variant. While that is helpful, it also means that each module people want to generate will result in a separate installations of the `intel-oneapi-compilers` package.

I think it might be preferable to create a package which depends on `intel-oneapi-compilers` called something like `intel-oneapi-icc-wrappers`, and *that* package can create a module which sets `CC=icc` etc. - does that seem reasonable?

Side note: This seems to relate to `use_variant`, but IMO, a user who explicitly wants to access these different compiler executables may in fact benefit from an explicit `-wrapper`-style package.

rscohn2 replied:
@mathomp4: What do you think about @scheibelp suggestion. I don't use module files and do not know if it is important that the package name be intel-oneapi-compilers

mathomp4 replied:
@rscohn2 Hmm. I don't know if it would matter too much. For my application that I'm hoping to bring into spack "soon" (I have to submit about 6 new package PRs first), just a few `module load` commands bring in like 30-40 modules as LMod does its dependency magic.

So probably in the end, I'd present to users a sort of "meta-module" that loads everything underneath. I could call *that* module anything in regards to the compiler name as long as I know what the "real" intel module is called.

rscohn2 committed:
lmod needs to set family, remove variant

rscohn2 committed:
split out classic package for wrappers, add lua support

rscohn2 commented:
To address @scheibelp comments, I split classic compilers into a separate package. The classic compilers are going away so I wanted to make intel-oneapi-compilers work for people who want icx. icc users can use intel-oneapi-compilers-classic if they want modules to be set up for that compiler.

I also addressed an issue from #30608 with a fix from https://github.com/spack/spack/issues/24195#issuecomment-856576338, so asking @ilbiondo to look at this. I do not use modules and did limited testing.

rscohn2 commented:
@scheibelp: All required tests are passing. Can you review this?

scheibelp commented on var/spack/repos/builtin/packages/intel-oneapi-compilers-classic/package.py:
```diff
@@ -0,0 +1,50 @@
+                '2021.2.0',
+                '2021.1.2']:
+        version(ver)
+        depends_on('intel-oneapi-compilers@' + ver, when='@' + ver)
```
This should also set `type` so that the `intel-oneapi-compilers` `bin/` directory is added to PATH:

```
depends_on('intel-oneapi-compilers@' + ver, when='@' + ver, type='run')
```

scheibelp reviewed (changes requested):
I have one minor request.

Overall this looks good but I want to confirm:  Users would install `intel-oneapi-compilers`, `intel-oneapi-compilers-classic`, and/or `dpcpp` to generate the modules they want? This looks like it would work but is a slight departure from the PR as originally written since `dpcpp` wouldn't be using the `intel-oneapi-compilers` installation (which IIRC comes with `dpcpp`). As long as that's expected/desired that works with me.

rscohn2 committed:
add run to depends_on

rscohn2 commented:
> Overall this looks good but I want to confirm: Users would install `intel-oneapi-compilers`, `intel-oneapi-compilers-classic`, and/or `dpcpp` to generate the modules they want?

dpcpp package is for people who want to use the open source compiler. I could make a intel-oneapi-compilers-dpcpp package, but for most users, 'icpx -fsycl' is the same as dpcpp so I did not want the extra package.

> This should also set `type` so that the `intel-oneapi-compilers` `bin/` directory is added to PATH:

Added

rscohn2 commented:
@scheibelp : Can you merge? You approved and all required tests have passed.

bryank-cs commented:
There's still a problem with this fix. Although it correctly sets the `MODULEPATH`, because you have that `intel-oneapi-compilers-classic` depends on `intel-oneapi-compilers`, and they both provide a compiler, then when `lmod` tries to load the dependency, it sees a second `family("compiler")` and unloads the `-classic` version:
```
⮞ module load intel-oneapi-compilers-classic

Lmod is automatically replacing "intel-oneapi-compilers-classic/2021.7.0-tunk4y4" with "intel-oneapi-compilers/2022.2.0-u53gtvl".
```

rscohn2 commented:
I don't see that problem. @skosukhin: Do you know what could be the cause?

```
rscohn1@anpfclxlin02:spack$ source spack/share/spack/setup-env.sh
rscohn1@anpfclxlin02:spack$ module avail

------------------ /localdisk/work/rscohn1/projects/spack/spack/share/spack/modules/linux-ubuntu20.04-cascadelake -------------------
   intel-oneapi-compilers-2022.2.0-gcc-10.3.0-ks33nah            patchelf-0.15.0-gcc-10.3.0-zqhohe7
   intel-oneapi-compilers-classic-2021.7.0-gcc-10.3.0-cnylmbw

------------------------------------------------- /usr/share/lmod/lmod/modulefiles --------------------------------------------------
   Core/lmod/6.6    Core/settarg/6.6

Use "module spider" to find all possible modules.
Use "module keyword key1 key2 ..." to search for all possible modules matching any of the "keys".

rscohn1@anpfclxlin02:spack$ module load intel-oneapi-compilers-classic-2021.7.0-gcc-10.3.0-cnylmbw
rscohn1@anpfclxlin02:spack$ icc --version
icc: remark #10441: The Intel(R) C++ Compiler Classic (ICC) is deprecated and will be removed from product release in the second half of 2023. The Intel(R) oneAPI DPC++/C++ Compiler (ICX) is the recommended compiler moving forward. Please transition to use this compiler. Use '-diag-disable=10441' to disable this message.
icc (ICC) 2021.7.0 20220726
Copyright (C) 1985-2022 Intel Corporation.  All rights reserved.

rscohn1@anpfclxlin02:spack$ which icc
/localdisk/work/rscohn1/projects/spack/spack/opt/spack/linux-ubuntu20.04-cascadelake/gcc-10.3.0/intel-oneapi-compilers-2022.2.0-ks33nah3lgcbvkmrey66fhgv4adz5tnl/compiler/2022.2.0/linux/bin/intel64/icc
rscohn1@anpfclxlin02:spack$
```

rscohn2 commented:
There was a defect in packaging in the 2023.2.0 compiler, so they made a 2023.2.1 release with the same 2023.2.0 binary and spack gets the compiler version number from the binary.  I did not know at the time that intel had done this. I think that was a 1 time thing, but the intel-oneapi-compilers also contains %intel toolchain, and they have their own version numbers (intel@2021.10.0). I don't know if there is a good way to represent this in spack. I believe this was the last release of the %intel toolchain so going forward, the package version and compiler version will be in sync and we won't have this issue.
